### PR TITLE
Add π0 VLA model inference support

### PR DESCRIPTION
**Pull request by yicwang**
## Motivation

Issue: https://github.com/sgl-project/sglang/issues/18266. This PR adds first-class inference support for **[π0 (Pi-Zero)](https://www.physicalintelligence.company/blog/pi0)**, the Vision-Language-Action (VLA) model from Physical Intelligence, to SGLang. π0 takes multi-camera images + a language instruction + robot proprioceptive state and outputs a **continuous action chunk** via flow-matching denoising — it does not emit text tokens. Reference implementations live in [openpi](https://github.com/Physical-Intelligence/openpi) and [lerobot](https://github.com/huggingface/lerobot).

Bringing π0 under SGLang gets the robotics community:
- a single production-grade serving stack (HTTP / OpenAI-compatible API, request queueing, metrics, weight updates, etc.) instead of a one-off Flask-like policy server;
- the SGLang multimodal input pipeline (URL / base64 / PIL / precomputed) reused unchanged;
- a clean extension point for future VLA models (π0.5, π0-FAST, SmolVLA, etc.) — most of the plumbing in this PR is model-agnostic.

Parity against LeRobot's `PI0Policy` is bit-for-bit: `max|Δ| < 2e-6` on the action chunk for `lerobot/pi0_base` with fixed noise (see *Accuracy Tests* below).

## Modifications

This PR is intentionally split into **6 logically independent commits** so each one can be reviewed in isolation:

1. **`[pi0] config: register Pi0Config and VLA / multimodal architecture tags`**
   - New `sglang.srt.configs.pi0.Pi0Config` (HF-compatible): paligemma/action-expert variants, chunk/action/state dims, flow-matching `num_inference_steps`, square-only `image_resolution` (fail-fast validated), optional `norm_stats` block, and plumbing fields for `ModelConfig._derive_model_shapes`.
   - Register `Pi0Config` in `_CONFIG_REGISTRY` (in the new `hf_transformers/common.py`) so `model_type: "pi0"` resolves before `sglang.srt.models.pi0` is imported.
   - Register `Pi0ForActionPrediction` in **both** `is_vla_model()` **and** `multimodal_model_archs` — π0's HF config has no nested `vision_config` (SigLIP dims come from `paligemma_variant`), so we rely on the multimodal list to route startup through `get_processor` / `get_mm_processor`.
   - `hf_transformers/processor.py::get_processor` routes pi0 through a **tokenizer-only fast path** (`AutoTokenizer` instead of `AutoProcessor`) — the `lerobot/pi0_base` checkpoint ships no processor_config.

2. **`[pi0] model: add π0 VLA (PaliGemma + action expert + flow matching)`**
   Inference-only `Pi0ForActionPrediction`:
   - **PaliGemma 2B** backbone (SigLIP vision + Gemma 2B LM) produces prefix tokens.
   - **Gemma 300M action expert** consumes `[state, noisy-action, time]` suffix tokens.
   - **Flow-matching head**: iterative Euler denoising from noise at `t=1` to actions at `t=0`.
   - Targets **modern transformers** (no dependency on OpenPI's `transformers_replace` monkey-patch). We walk Gemma decoder layers manually through `_compute_layer_*` helpers and keep a plain `list[(k, v)]` KV cache — sidesteps the SDPA/eager attention-mask dance and HF `DynamicCache` churn.
   - LeRobot-compatible state/action (un)normalization (optional, passthrough when `norm_stats` not set).
   - `load_weights()` remaps the `lerobot/pi0_base` checkpoint keys (flat → nested PaliGemma layout, `paligemma.lm_head.weight` → `embed_tokens.weight` for the tied embedding).
   - Uses the transformers v5 `dtype` kwarg (not deprecated `torch_dtype`).
   - Warmup / CUDA-graph capture compat: logs a WARNING + returns zeros when `mm_input` is missing, with a one-shot flag so it doesn't flood the log on real traffic.

3. **`[pi0] multimodal processor: images + state + language tokenization`**
   `Pi0Processor` produces the per-request tensors `Pi0ForActionPrediction.forward` consumes:
   - `pixel_values (num_cameras, 3, 224, 224)` padded with `-1` for unused cameras;
   - `image_masks (num_cameras,) bool`;
   - `lang_tokens (L,)` long — PaliGemma tokenizer, `\n`-terminated (LeRobot's `Pi0NewLineProcessor` semantics), right-padded to `max_token_len`;
   - `state (state_dim,) float32` (optional);
   - `num_inference_steps` override (optional).
   - Image preprocessing (aspect-preserving resize + `-1` pad + SigLIP normalization) matches OpenPI/LeRobot bit-for-bit.
   - Returns a typed `MultimodalProcessorOutput` + sets `MultimodalDataItem.offsets` so `pad_input_tokens` stamps the per-item pad value onto the image-placeholder run.
   - Rejects `>3` cameras (π0 is fixed at 3 camera slots).
   - **Bypasses `load_mm_data()`**: π0's prompt has no text-level `<image>` placeholders (placeholders are positional, not textual), so the generic text-scan warning is irrelevant and we skip it.
   - `_process_precomputed()` fast path for clients that already have normalized pixel values, with a strict contract (single stacked `(num_cameras, C, H, W)` feature + validated metadata) so downstream stays identical to the normal path.

4. **`[pi0] runtime: VLA batch result plumbing across scheduler + IPC`**
   VLA models produce continuous action vectors during prefill — no decode loop, no sampling. This commit adds a typed VLA result path across scheduler + IPC + detokenizer + tokenizer manager:
   - New `VLABatchResult` dataclass (`scheduler.py`) and `FINISH_VLA` finish reason (`schedule_batch.py`, serializes to `{"type": "stop", "matched": "vla_done"}`).
   - `Scheduler.is_vla` mirrors `ModelConfig.is_vla`; `run_batch()` short-circuits on the VLA **extend** branch — runs the model, nulls `batch_result.next_token_ids` (so no downstream reader consumes `argmax` over action floats), populates a dummy `batch.output_ids` so overlap scheduling / future-resolution doesn't crash on the next iteration.
   - `is_extend()` guard is load-bearing: IDLE / DECODE / prebuilt batches fall through to the generic path so captured decode-mode graphs replay unchanged.
   - `BatchTokenIDOutput` / `BatchStrOutput` gain a typed `vla_actions` field — passed through `detokenizer_manager` unchanged, then promoted to top-level `actions` by a new `_promote_vla_actions()` helper in `tokenizer_manager`.
   - `process_batch_result_vla` finishes every request with `FINISH_VLA`, attaches the chunk on `req.vla_actions`, and releases the KV pool slot only if one was allocated.
   - `event_loop_overlap` skips `launch_batch_sample_if_needed` for VLA (no `delay_sample_func`).
   - `scheduler_runtime_checker_mixin`: VLA-specific waiver on busy/idle KV-pool leak asserts — see *Limitations* below for why.

5. **`[pi0] tests: CPU-only unit tests + LeRobot parity test`**
   - `test/srt/models/test_pi0.py` (CPU-only, no weights, runs in seconds): sinusoidal timestep embedding, `make_att_2d_masks` semantics (bidirectional / causal / prefix-LM / padding), `prepare_attention_masks_4d`, `_build_norm_buffers` / `_apply_norm` round-trip (incl. padded-tail preservation), `Pi0Config` defaults + non-square resolution rejection, `GemmaVariantConfig`, `is_vla_model` + `is_multimodal_model` registration, image preprocessing (resize / RGBA→RGB), `Pi0Processor` helpers, `load_weights` remap rules (flat → nested, `lm_head` → `embed_tokens`), `_promote_vla_actions` helper, `FINISH_VLA.to_json()` contract, and a regression grep that the scheduler VLA branch still nulls `next_token_ids`.
   - `test/srt/models/test_pi0_parity.py` (opt-in, requires weights): instantiates LeRobot `PI0Policy` and SGLang `Pi0ForActionPrediction` from the **same** `lerobot/pi0_base` checkpoint, feeds **identical** pre-processed inputs and **fixed** noise, asserts `torch.allclose` on the action chunk. Optional OpenPI comparison behind `PI0_PARITY_CHECK_OPENPI=1`. Per-stage diagnostics (prefix embeddings / layer-0 KV / first denoise v_t) print on mismatch. Skipped in CI.

6. **`[pi0] docs: add supported-models page for π0 VLA`**
   `docs/supported_models/pi0_vla.md` covers: setup (HF weights download, `config.json` patch, pulling the PaliGemma tokenizer with `add_bos_token: true`, launch command), example `/generate` and OpenAI-compatible chat.completions requests passing `state` + `num_inference_steps` via `extra_body`, response schema (top-level `actions`, `FINISH_VLA` matched="vla_done"), normalization contract (client-side default vs server-side via `norm_stats`), inference-pipeline walkthrough + attention mask diagram, and references.

### Current scheduler-level limitations (and the planned follow-ups)

This PR deliberately keeps the scheduler change surface small — the denoise loop is run **inside** a single `extend` forward rather than as a new scheduler phase. That keeps this PR focused on *"VLA inference correctness"* but leaves several optimizations on the table. The natural follow-up PRs are:

- **Step-level batching across requests.** Today `Pi0.forward` runs the full `num_inference_steps` (10 by default) Euler loop inline inside one `extend` forward, and the per-request `for` loop inside it is Python, not vectorized on the batch dim. Across multiple concurrent requests there is no continuous-batching / step interleaving. *Follow-up: promote each denoise step to a scheduler-level micro-batch and vectorize the batch dim inside `sample_actions`.*

- **Prefix KV cache reuse.** The PaliGemma prefix KV cache is a `list[(k, v)]` living on the Python stack inside `Pi0.forward` — it is not in `token_to_kv_pool_allocator` and not indexed by RadixAttention. Same `(prompt, images)` pair gets recomputed on every request. *Follow-up: plumb the prefix KV through the shared KV pool so RadixAttention hits identical prefixes; this also lets us drop the VLA waiver on the KV-pool leak check.*

- **Attention backend.** We use a hand-rolled `_attend` (eager softmax) with a custom `(B, 1, Q, prefix+suffix)` mask slice. It's correct but not FlashInfer/Triton-fast. *Follow-up: port to a mask-aware FlashInfer / Triton kernel. π0's mask is non-causal within prefix and prefix-LM for suffix, so it needs a custom mask kernel or a two-pass variant.*

- **CUDA graph capture.** VLA currently runs eager. We can't just set `is_prefill_only=True` without breaking CUDA-graph replay (the extend path would get routed through a decode-mode graph that expects initialized `sampling_info.temperatures`), and warmup can legitimately reach `Pi0.forward` with no `mm_input` (we handle this by logging + returning zeros). *Follow-up: capture prefix and per-step denoise as their own graphs.*

- **Overlap scheduling.** `VLABatchResult` has no `copy_to_cpu` / `copy_done` and `event_loop_overlap` skips `launch_batch_sample_if_needed` for VLA — so overlap scheduling is currently a no-op for VLA. *Follow-up: re-enable once the above lands, so the denoise loop can overlap with the next request's prefix forward.*

- **Runtime-checker waiver.** The KV-pool leak asserts (busy + idle) are currently skipped globally for any `is_vla` model. This is safe for π0 specifically (its real memory — the PaliGemma KV cache — is private torch tensors in its own forward), but it's wider than it needs to be. *Follow-up: narrow to a per-arch condition, or drop entirely once VLA uses the shared KV pool.*

None of the above blocks correctness; calling them out so reviewers know exactly what's out of scope for this PR and what the next PRs will look like.

## Accuracy Tests

**Parity vs LeRobot `PI0Policy`** — same `lerobot/pi0_base` weights, same preprocessed inputs, same fixed noise:

```bash
PYTHONPATH=openpi/src:lerobot/src:sglang/python \
    python -m pytest sglang/test/srt/models/test_pi0_parity.py -v -s
[parity] state.shape=torch.Size([2, 32])  lang_tokens.shape=torch.Size([2, 48])
[parity] images[0].shape=torch.Size([2, 3, 224, 224]) (num_cams=3)
[parity] noise.shape=torch.Size([2, 50, 32])  dtype=torch.float32
[parity] |Δ| max=1.43e-06  mean=2.7e-07  atol=1.0e-04
[parity] torch.allclose(atol=1e-4): True
```

Also verified against OpenPI's reference `PI0Pytorch` with `PI0_PARITY_CHECK_OPENPI=1` — same result (`max|Δ| < 2e-6`).
CPU-only unit tests (39 tests, no weights required, seconds):

```bash
python -m pytest sglang/test/srt/models/test_pi0.py -v
----------------------------------------------------------------------
Ran 50 tests in 87.212s

OK
```

## Speed Tests and Profiling

Not applicable for this PR — π0 is a new model, not a perf change to an existing one. Wall-clock for a single request on L4 fp32 is ~100 ms (1 PaliGemma prefill + 10 denoise steps). The *scheduler-level limitations* section above lists the follow-ups that would actually make this competitive with SGLang's decode path on throughput; I'd prefer to land those as separate PRs with their own benchmarks.

## Checklist

- Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed). (Speed benchmark N/A — see above.)
- Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** — add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** — `run-ci` is required first.

### CODE REVIEW

**Comment by 1e4ves** [comment]
I reproduced the PR and found one issue that is not covered by the current parity test.

  `test_pi0_parity.py` validates the model-level path by directly calling `sglang_model.sample_actions(...)`, so it bypasses the actual `/generate` server path. 
  In the server path,    `Pi0ForActionPrediction.forward()` packs the continuous VLA action chunk into`logits_output.next_token_logits`, but `forward_batch_generation()` then runs the normal token sampler. The sampler mutates `next_token_logits` in-place via softmax, so `process_batch_result_vla()` later reads softmax probabilities as actions.

  Observed e2e before fix:

  ```text
  server_mean ≈ 0.000625
  server_std ≈ 0.00018
  reference_mean ≈ -0.05359
  reference_std ≈ 0.29939
  max_diff ≈ 0.9816
  mean_diff ≈ 0.2230
```
  After bypassing the sampler for VLA extend batches, /generate matches the LeRobot reference:
```text
  server_mean = -0.0535949
  server_std = 0.299394
  reference_mean = -0.0535949
  reference_std = 0.299394
  max_diff = 2.98e-07
  mean_diff = 6.30e-08
  allclose(atol=1e-4) = True
```
  I also noticed that the docs use extra_body.state and extra_body.num_inference_steps for
  raw /generate, but GenerateReqInput currently does not preserve extra_body, and
  GenerateReqInput.__getitem__() does not pass it to per-request objects. That means these
  fields can be dropped before the π0 processor reads them.

  Would you prefer a patch pushed into this PR branch, or a separate follow-up PR?

**Comment by yicwang** [comment]
> I reproduced the PR and found one issue that is not covered by the current parity test.
> 
> `test_pi0_parity.py` validates the model-level path by directly calling `sglang_model.sample_actions(...)`, so it bypasses the actual `/generate` server path. In the server path, `Pi0ForActionPrediction.forward()` packs the continuous VLA action chunk into`logits_output.next_token_logits`, but `forward_batch_generation()` then runs the normal token sampler. The sampler mutates `next_token_logits` in-place via softmax, so `process_batch_result_vla()` later reads softmax probabilities as actions.
> 
> Observed e2e before fix:
> 
> ```
> server_mean ≈ 0.000625
> server_std ≈ 0.00018
> reference_mean ≈ -0.05359
> reference_std ≈ 0.29939
> max_diff ≈ 0.9816
> mean_diff ≈ 0.2230
> ```
> 
> After bypassing the sampler for VLA extend batches, /generate matches the LeRobot reference:
> 
> ```
>   server_mean = -0.0535949
>   server_std = 0.299394
>   reference_mean = -0.0535949
>   reference_std = 0.299394
>   max_diff = 2.98e-07
>   mean_diff = 6.30e-08
>   allclose(atol=1e-4) = True
> ```
> 
> I also noticed that the docs use extra_body.state and extra_body.num_inference_steps for raw /generate, but GenerateReqInput currently does not preserve extra_body, and GenerateReqInput.**getitem**() does not pass it to per-request objects. That means these fields can be dropped before the π0 processor reads them.
> 
> Would you prefer a patch pushed into this PR branch, or a separate follow-up PR?

I intended to bypass the /generate path for parity test, because it is much faster for debugging and should be OK. But what you pointed out it is indeed a bug. I just fixed it and now the document is accurate now.

Just rebased and pushed out the latest.

File: docs/supported_models/pi0_vla.md
Comment (zijiexia) [inline docs/supported_models/pi0_vla.md:1]: Hi @yicwang , we've migrated our docs to the new location under docs_new. Could you please migrate your changes to docs_new accordingly? Thanks!

**Review by zijiexia — CHANGES_REQUESTED** [review]

File: docs/supported_models/pi0_vla.md
Comment (yicwang) [reply docs/supported_models/pi0_vla.md:1]: Done
